### DISCUSSION TITLE
Changing fonts in different plot types

### DISCUSSION THREAD

**Discussion by raagagrawal** [uclahs-cds/public-R-BoutrosLab-plotting-general#180] [Q&A]
I am looking to change the font family used across a number of plot types. Below is a toy example of a barplot I am making that I would like to change the font family of to 'Arial'.
```

library(BoutrosLab.plotting.general)

set.seed(123)  # For reproducibility

# Define the groups, grades, and stains
groups <- c("Group A", "Group B", "Group C")
grades <- c("Positive")
stains <- c("Stain 1", "Stain 2")

# Generate random frequencies
frequencies <- sample(1:20, 6, replace = TRUE)

# Function to calculate percentages
calculate_percentage <- function(frequencies) {
  total <- sum(frequencies)
  return(frequencies / total * 100)
}

# Assign frequencies and calculate percentages for each stain
freq_stain1 <- frequencies[1:3]
freq_stain2 <- frequencies[4:6]
percentages_stain1 <- calculate_percentage(freq_stain1)
percentages_stain2 <- calculate_percentage(freq_stain2)

# Create the data frame
combined_result_df <- data.frame(
  Group = rep(groups, 2),
  Grade = rep(grades, each = 3, times = 2),
  Frequency = c(freq_stain1, freq_stain2),
  Percentage = c(percentages_stain1, percentages_stain2),
  Stain = rep(stains, each = 3)
)

# Create bar plot
barplot <- create.barplot(
  formula = Percentage ~ Group | Stain,
  data = combined_result_df,
  main = NULL,
  ylab.lab = 'Positive (%)',
  xlab.label = NULL,
  xaxis.lab = rep('', length(groups)),
  ylimits = c(0, 100),
  yat = seq(0, 100, 25),
  xaxis.cex = 1,
  yaxis.cex = 1,
  xlab.cex = 0,
  ylab.cex = 1.5,
  xaxis.fontface = 1,
  yaxis.fontface = 1,
  col = c('Black'),
  strip.cex = 0,
  xaxis.tck = 0,
  yaxis.tck = 0,
  resolution = 100
)

barplot
```

A previously suggested method searched the structure of the plot to find all occurrences of 'fontfamily' and replaced it with our desired font family. However, this method does not work in this plot example.

I have had minor success updating the plot using update.trellis...however this is inconsistent and the font will change when written to a png.

```

# Function to update font to Arial and save plot as PNG
update_and_save_plot <- function(plot, filename) {
  # Ensure the Arial font is used in the plot settings
  plot <- update(plot, par.settings = list(
    axis.text = list(family = "Arial"),
    par.main.text = list(family = "Arial"),
    par.xlab.text = list(family = "Arial"),
    par.ylab.text = list(family = "Arial"),
    add.text = list(family = "Arial"),
    par.sub.text = list(family = "Arial"),
    y.scales = list(family = "Arial")
  ))
  # Save the plot as a PNG file with cairo to handle fonts better
  png(filename, res = 500, width = 5, height = 2, units = "in", type = 'cairo')
  print(plot)
  dev.off()
}

update_and_save_plot(barplot, filename)
```

I would appreciate any and all ideas! Thank you in advance!

**Comment by jmlivingstone** [comment]
I am not sure why but the default fontfamily was removed from BPG in this commit
https://github.com/uclahs-cds/package-BoutrosLab-plotting-general/commit/27cce390f33458fdcc8e19ca7330626011ce16f9
Now when create.* calls BoutrosLab.plotting.general::get.defaults(property = 'fontfamily'), NULL is returned. Stefan's example will no longer work because fontfamily is not being set so it won't find anything to replace.

I was just able to set fontfamily in the parameters I wanted to change like this

```
plot <- create.barplot(
  formula = Percentage ~ Group | Stain,
  data = combined_result_df,
  file = 'with_default_font.png', 
  main = 'Title',
  ylab.lab = 'Positive (%)',
  xaxis.lab = rep('', length(groups)),
  ylimits = c(0, 100),
  yat = seq(0, 100, 25),
  xaxis.cex = 1,
  yaxis.cex = 1,
  xlab.cex = 1.5,
  ylab.cex = 1.5,
  xaxis.fontface = 1,
  yaxis.fontface = 1,
  col = c('Black'),
  strip.cex = 0,
  xaxis.tck = 0,
  yaxis.tck = 0,
  resolution = 100
  )

plot$main$fontfamily <- 'Mono'
plot$ylab$fontfamily <- 'Arial'
plot$xlab$fontfamily <- 'Garamond'

png('with_font_changed.png')
plot
dev.off()
```
![with_font_changed](https://github.com/uclahs-cds/package-BoutrosLab-plotting-general/assets/4053843/0384d3b2-1945-42f1-a62b-de877ba04fcd)

**Reply by raagagrawal** [reply]
Thank you @jmlivingstone for this solution! This definitely lets me change the font. But I think the Arial font here is actually not Arial. Look at the shape of the 't'. Do you know what could be happening?
![image](https://github.com/uclahs-cds/package-BoutrosLab-plotting-general/assets/10327284/1126afe8-53ad-4897-a5f9-8e4c3a39c378)

**Reply by raagagrawal** [reply]
On further review...the Arial on the cluster I think is actually just MS Sans Serif...http://www.identifont.com/differences?first=Microsoft+Sans+Serif&second=Arial

**Reply by raagagrawal** [reply]
Were you able to change the fontfamily for the y.scale? It seems to not change to Arial (look at the 7..there should be a slight curve).
